### PR TITLE
Multiplayer Desync Debug Logging

### PR DESCRIPTION
**Pull request by AMZN-puvvadar**
Initial version of Desync Debug Logging accessible through the Debug Dropdown Menu (accessible with the Home key by default)
Includes:
- Logging of all desyncs
- Logging of inputs with non default values
- Logging of values specified by the user via macro
- ImGUI Dialog displaying logs with search and refresh functionality

The course of the work showed a general need for string converters for math types so these were moved into AzCore from EMFX. bool was specifically left alone as specifying it in AzCore seems to break AP.

Tests were not included for changes implemented in Multiplayer.Debug. Some tests were added for general MP changes. Linux has not been tested due to the lack of a machine.

Signed-off-by: puvvadar <puvvadar@amazon.com>

### CODE REVIEW

File: Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h [hunk 1]
```diff
@@ -7,9 +7,177 @@
  */
 
 #pragma once
+#include <AzCore/Math/MathStringConversions.h>
 
 namespace Multiplayer
 {
+    enum class MultiplayerAuditCategory
+    {
+        MP_AUDIT_DESYNC,
```
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h:16]: Since this is an enum class, and you have to scope usage in any case, you can go with more readable names:

```
enum class AuditCategory
{
   Desync, 
   Input,
   Event
}
```

and then you have `Multiplayer::AuditCategory::Input` etc.. rather than triplicate the multiplayer bit, duplicate the audit bit, and use macro name formatting, `Multiplayer::MultiplayerAuditCategory::MP_AUDIT_INPUT`

File: Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h [hunk 2]
```diff
@@ -7,9 +7,177 @@
...
+        MP_AUDIT_INPUT,
+        MP_AUDIT_EVENT
+    };
+
+    //! @class IMultiplayerAuditingDatum
+    //! @brief IMultiplayerAuditingDatum provides an interface for auditing datums of a given auditing event
+    class IMultiplayerAuditingDatum
```
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h:23]: Try and move all your impls to an inline file, it's really difficult to see what your interfaces actually are (and there are no docs markup yet as well) with all the implementations mixed in together like this.

Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h:23]: Good call, I've cleaned this up a fair bit.

**Comment by lmbr-pip** [comment]
Added sig/core as PR involves new string conversions from EMFX to Code/Framework/AzCore/AzCore/Math/MathStringConversions.cpp files

File: Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h [hunk 3]
```diff
@@ -7,9 +7,64 @@
...
+                elements.resize(rhs.elements.size());
+                for (int32_t i = 0; i < rhs.elements.size(); ++i)
+                {
+                    elements[i] = rhs.elements[i]->Clone();
+                }
+            }
+            return *this;
+        }
+
+        AZStd::string name;
```
Comment (lmbr-pip) [inline Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h:64]: Why break encapsulation by exposing member variables? These should also be m_ as per C++ coding standards (https://github.com/o3de/sig-core/blob/main/governance/Coding-Standards-and-Style-Guide.md#variable-names)

File: Gems/Multiplayer/Code/Include/Multiplayer/MultiplayerDebug.h [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+#include <AzCore/Math/MathStringConversions.h>
+#include <Multiplayer/IMultiplayerDebug.h>
+
+namespace Multiplayer
+{
+    template<class T>
+    class MultiplayerAuditingDatum : public IMultiplayerAuditingDatum
+    {
+    private:
```
Comment (lmbr-pip) [inline Gems/Multiplayer/Code/Include/Multiplayer/MultiplayerDebug.h:19]: nit: Consider public things first rather than private. As a user I care about what I can call, not so much the implementation

File: Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp [hunk 5]
```diff
@@ -189,15 +217,24 @@ namespace Multiplayer
 
                 if (lostInput)
                 {
+                    //TODO: Maybe add to Audit Trail here (server)
```
Comment (lmbr-pip) [inline Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp:220]: Consider migrating TODOs to public GitHub issues. TODOs are hard for other developers to take action on as its unclear who is doing the work or why

File: Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp [hunk 6]
```diff
@@ -312,30 +352,39 @@ namespace Multiplayer
         SerializeEntityCorrection(serializer);
         GetNetBindComponent()->NotifyCorrection();
 
-#ifndef AZ_RELEASE_BUILD
+        const uint32_t inputHistorySize = static_cast<uint32_t>(m_inputHistory.Size());
+        const uint32_t historicalDelta = aznumeric_cast<uint32_t>(m_clientInputId - inputId); // Do not replay the move we just corrected, that was already processed by the server
```
Comment (lmbr-pip) [inline Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp:356]: nit: avoid personal  pronouns in comments 'we' is often unclear to the reader. Also as this code can be run in game, server or editor again, the we is unclear.

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.h [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        AZStd::string_view GetAuditTrialFilter();
+
+        //! Sets string filter for the audit trail
+        void SetAuditTrailFilter(AZStd::string_view filter);
+    private:
+        AZ::ScheduledEvent m_updateDebugOverlay;
+
+        AZStd::string_view m_filter;
+        AzFramework::DebugDisplayRequests* m_debugDisplay = nullptr;
+        char m_inputBuffer[1028] = {};
```
Comment (lmbr-pip) [inline Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.h:69]: Consider moving magic number to const or similar and include how set. Also note fixed sized 'dumb' arrays have historically caused a lot of pain points around  overflow so be careful that use of buffer is appropriately guarded.

**Review by lmbr-pip — COMMENTED** [review]
Can we update description as it seems out of date?

Also be add:
* How tested?
* If you tested on Linux, as its now reference platform?
* Reason for no automated tests (unit, integration) with code.

File: Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h (hunk 3 above)
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h:64]: This really should be a struct as it's an abstract value data container. Will fix the naming as well.

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.h (hunk 7 above)
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.h:69]: From what I'm understanding of ImGui (the user of this buffer), it should be enforced. I'll add a const.

File: Code/Framework/AzCore/AzCore/Math/MathStringConversions.h [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    class Transform;
+    class Vector2;
+    class Vector3;
+    class Vector4;
+}
+
+namespace AZStd
+{
+    //! Prints a Vector2 with precision to 8 decimal points
+    void to_string(string& str, const AZ::Vector2& value);
```
Comment (lmbr-pip) [inline Code/Framework/AzCore/AzCore/Math/MathStringConversions.h:28]: I believe these names should be camel case as per core coding standards: https://github.com/o3de/sig-core/blob/main/governance/Coding-Standards-and-Style-Guide.md

File: Gems/EMotionFX/Code/MCore/Source/StringConversions.h [hunk 9]
```diff
@@ -50,18 +41,6 @@ namespace MCore
 namespace AZStd
 {
     void to_string(string& str, bool value);
```
Comment (lmbr-pip) [inline Gems/EMotionFX/Code/MCore/Source/StringConversions.h:43]: nit: Why not lift all the conversions out?

File: Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h [hunk 10]
```diff
@@ -7,9 +7,70 @@
...
+        virtual AZStd::unique_ptr<IMultiplayerAuditingDatum> Clone() = 0;
+    };
+
+    //! @struct MultiplayerAuditingElement
+    //! @brief MultiplayerAuditingElement contains a list of datums for a given auditing event
+    struct MultiplayerAuditingElement
+    {
+    public:
+        MultiplayerAuditingElement() = default;
+        MultiplayerAuditingElement(const MultiplayerAuditingElement& rhs)
```
Comment (lmbr-pip) [inline Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h:46]: Did we consider move assignment operator for struct to avoid deep clones? or is there always a reason for deep clones?

File: Gems/Multiplayer/Code/Include/Multiplayer/NetworkInput/NetworkInput.h [hunk 11]
```diff
@@ -52,6 +53,11 @@ namespace Multiplayer
 
         bool Serialize(AzNetworking::ISerializer& serializer);
 
+        void LogComponentInputDelta() const;
```
Comment (lmbr-pip) [inline Gems/Multiplayer/Code/Include/Multiplayer/NetworkInput/NetworkInput.h:56]: nit: consider autogen comments on all public headers

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.h [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        void UpdateDebugOverlay();
+
+        //! Checks if the audit trail can be pumped and resets the pump flag
+        bool CanPumpAuditTrail();
+
+        //! Gets string filter for the audit trail
+        AZStd::string_view GetAuditTrialFilter();
+
+        //! Sets string filter for the audit trail
+        void SetAuditTrailFilter(AZStd::string_view filter);
```
Comment (lmbr-pip) [inline Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.h:66]: nit: pass as const ref?

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.h [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    public:
+        static constexpr char DESYNC_TITLE[] = "Desync on %s";
+        static constexpr char INPUT_TITLE[] = "%s Inputs";
+        static constexpr char EVENT_TITLE[] = "DevEvent on %s";
+
+        MultiplayerDebugAuditTrail();
+        ~MultiplayerDebugAuditTrail();
+
+        //! Main update loop.
+        void OnImGuiUpdate(AZStd::deque<AuditTrailInput> auditTrailElems);
```
Comment (lmbr-pip) [inline Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.h:54]: nit: Why pass by value?

File: Gems/Multiplayer/Code/Source/NetworkInput/NetworkInput.cpp [hunk 14]
```diff
@@ -86,6 +86,39 @@ namespace Multiplayer
         return m_hostBlendFactor;
     }
 
+    void NetworkInput::LogComponentInputDelta() const
+    {
+        for (uint16_t i = 0; i < m_componentInputs.size(); ++i)
+        {
+            //AZLOG_WARN(m_componentInputs[i].get()->GetInputDeltaLog().c_str());
```
Comment (lmbr-pip) [inline Gems/Multiplayer/Code/Source/NetworkInput/NetworkInput.cpp:93]: minor: Avoid committing commented out code. Consider if function has any value as is

File: Gems/Multiplayer/Code/Source/NetworkInput/NetworkInputChild.cpp [hunk 15]
```diff
@@ -25,6 +25,16 @@ namespace Multiplayer
...
 
+    bool NetworkInputChild::operator== (const NetworkInputChild& rhs) const
+    {
+        return m_owner == rhs.m_owner && m_networkInput.m_inputId == rhs.m_networkInput.m_inputId;
+    }
+
+    bool NetworkInputChild::operator!= (const NetworkInputChild& rhs) const
+    {
+        return !(m_owner == rhs.m_owner);
+    }
```
[the remark this answers is not in the document]
Comment (lmbr-pip) [reply Gems/Multiplayer/Code/Source/NetworkInput/NetworkInputChild.cpp:36]: Any updates on this? Is operator == still the right thing todo here?

**Review by lmbr-pip — COMMENTED** [review]
LG.

Pretty large PR. I would consider potentially breaking work up like this into smaller chunks in future, as theres a lot of details here.

File: Gems/EMotionFX/Code/MCore/Source/StringConversions.h (hunk 9 above)
Comment (AMZN-puvvadar) [reply Gems/EMotionFX/Code/MCore/Source/StringConversions.h:43]: The bool conversion specifically breaks AssetProcessor for reasons that are beyond me. I'm consequently keeping it scoped.

File: Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h (hunk 10 above)
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h:46]: The usage of unique_ptr in the elements vector (which allows it to function as an abstract value vector) forces it.

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.h (hunk 13 above)
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.h:54]: Looks like an oversight

File: Gems/Multiplayer/Code/Source/NetworkInput/NetworkInputChild.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/Source/NetworkInput/NetworkInputChild.cpp:36]: I believe so, though I'd welcome feedback from other sig-network contributors.

File: Code/Framework/AzCore/AzCore/Math/MathStringConversions.h (hunk 8 above)
Comment (AMZN-puvvadar) [reply Code/Framework/AzCore/AzCore/Math/MathStringConversions.h:28]: Which names? The variables are technically camel case by virtue of being single words and to_string while not, is the existing standard for the API.

File: Code/Framework/AzCore/AzCore/Math/MathStringConversions.cpp [hunk 16]
```diff
@@ -0,0 +1,71 @@
+/*
```
Comment (kberg-amzn) [inline Code/Framework/AzCore/AzCore/Math/MathStringConversions.cpp:1]: nice!

File: Code/Framework/AzCore/AzCore/Math/MathStringConversions.h (hunk 8 above)
Comment (kberg-amzn) [reply Code/Framework/AzCore/AzCore/Math/MathStringConversions.h:28]: AZStd follows the std naming conventions in all places for public methods. Hence Rajiv has picked the correct method names in this case.

File: Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h [hunk 17]
```diff
@@ -7,9 +7,70 @@
  */
 
 #pragma once
+#include <AzCore/Math/MathStringConversions.h>
 
 namespace Multiplayer
 {
+    //! Categories for Auditing logs used in MultiplayerDebugAuditTrail
+    enum class MultiplayerAuditCategory
```
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h:15]: You're still duping Multiplayer here, although it's much less of a big deal than before.. ie `Multiplayer::MultiplayerAuditCategory::Desync` instead of simply `Multiplayer::AuditCategory::Desync`

File: Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h [hunk 18]
```diff
@@ -7,9 +7,70 @@
...
+    //! @struct MultiplayerAuditingElement
+    //! @brief MultiplayerAuditingElement contains a list of datums for a given auditing event
+    struct MultiplayerAuditingElement
+    {
+    public:
+        MultiplayerAuditingElement() = default;
+        MultiplayerAuditingElement(const MultiplayerAuditingElement& rhs)
+        {
+            m_name = rhs.m_name;
+            m_elements.resize(rhs.m_elements.size());
```
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h:49]: resize_no_construct?

File: Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h [hunk 19]
```diff
@@ -24,5 +85,36 @@ namespace Multiplayer
...
+        //! Adds a string based entry to the Multiplayer audit trail
+        virtual void AddAuditEntry(
+            const MultiplayerAuditCategory category,
+            const ClientInputId inputId,
+            const HostFrameId frameId,
+            const AZStd::string& name,
+            AZStd::vector<MultiplayerAuditingElement>&& entryDetails) = 0;
     };
+
+    #define AZ_MPAUDIT_INPUT_REWINDABLE(INPUT, REWINDABLE, VALUE_TYPE)                                                            \
```
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h:98]: I would comment up these macros so everyone knows exactly what they do and what the parameters are.

File: Gems/Multiplayer/Code/Include/Multiplayer/NetworkTime/RewindableObject.h [hunk 20]
```diff
@@ -64,6 +64,10 @@ namespace Multiplayer
         //! @return value in const base type form
         const BASE_TYPE& GetPrevious() const;
 
+        //! Const base type retriever for value at last known authoritative update or the nearest to it.
+        //! @return value in const base type form
+        const BASE_TYPE& GetAuthority() const;
```
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/Include/Multiplayer/NetworkTime/RewindableObject.h:69]: This guy is a bit concerning. I think it might be worth having another look at. m_authTime is only updated on a serialize call, which is a bit undefined to be honest (could be a correction, could be a property update, etc..) and is zero at all other times. Also I don't know what the behaviour should be on a client vs a server.

Coupling to a role like authority at this level seems a bit wrong. You might want to back this up to 'GetLastSerializedValue()' or something a bit more generalized.

File: Gems/Multiplayer/Code/Source/NetworkInput/NetworkInputChild.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (kberg-amzn) [reply Gems/Multiplayer/Code/Source/NetworkInput/NetworkInputChild.cpp:36]: Yeah, this is fine.. it's required only for the delta tracking functionality so we can delta compress state updates through the serializer.

**Review by kberg-amzn — APPROVED** [review]
Geez you did a lot of good cleanup work on this. A few more comments, especially around the RewindableObject change that I really think should get looked at more carefully before submit, but overall really good job :thumbsup:

File: Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h (hunk 18 above)
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h:49]: Oh good call!

**Comment by AMZN-puvvadar** [comment]
Latest changes addressed some smaller feedback points and moves Auditing Macros from IMultiplayerDebug.h to MultiplayerDebug.h. This is to simplify the usage requirements of the macros as the template specifications in MultiplayerDebug.h are required to use the macros anyway.

File: Gems/Multiplayer/Code/Include/Multiplayer/NetworkTime/RewindableObject.h (hunk 20 above)
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/Include/Multiplayer/NetworkTime/RewindableObject.h:69]: I'll take a deeper look through this.

File: Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h (hunk 18 above)
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h:49]: I had to actually stick to resize. unique_ptr ends up attempting to invoke deleter on assignment so the lvalue seems to need to be constructed.

**Comment by AMZN-puvvadar** [comment]
Most recent change changes the authority value logic added to RewindableObject to last serialized value and adds a couple of guards around it. It also fixes a variety of exceptions (some of which that were in Debug mode only).

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugSystemComponent.cpp [hunk 21]
```diff
@@ -499,8 +542,145 @@ namespace Multiplayer
...
+                                        .contains(filter))
+                                {
+                                    filteredChild.m_elements.push_back(childElem.get()->Clone());
+                                }
+                            }
+
+                            if (filteredChild.m_elements.size() > 0)
+                            {
+
+                            filteredInput.m_children.push_back(filteredChild);
```
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugSystemComponent.cpp:671]: There's a bunch of weird formatting going on throughout these changes.. I'm not entirely sure what happened, but it's worth a quick pass to clean up anything that got a bit munged..

Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugSystemComponent.cpp]: Hmm yeah, this function has some oddities. I've cleaned them up.

**Comment by AMZN-puvvadar** [comment]
Fixed another IMGUI assert and cleaned up cvar usage. In particular, moved the Input auditing flag to LocalPredictionPlayerInputComponent where it more efficiently gates logic and got rid of the Event auditing flag entirely since it's an opt in feature.

**Comment by AMZN-puvvadar** [comment]
Latest set of changes adds a few things

- AZ_MPAUDIT_VALUE macro to let developers audit a variable when a NetworkInput isn't readily accessible, useful when several calls deep
- Updated Refresh logic for more consistency
- All Multiplayer Debug popups no longer let mouse clicks pass through to the game. In the Audit Trail in particular, this prevents event pollution when trying to refresh
- Added (limited) Desync events on the Server so the Audit Trail populates. This is useful when needing to compare Values across endpoints

File: Gems/EMotionFX/Code/MCore/Source/StringConversions.h (hunk 9 above)
Comment (burelc-amzn) [reply Gems/EMotionFX/Code/MCore/Source/StringConversions.h:43]: The bool conversion is gross. It adds an overload that has significant impact on overload resolution. Consider these
overloads:

```cpp
void to_string(AZStd::string& dest, AZStd::wstring_view src);
void to_string(string& str, bool value);
```

And then calling code like this:
```
WCHAR src[260];
AZStd::string dst;
AZStd::to_string(dst, src); // Which overload does this call?
```

If the .cpp has not included `MCore/Source/StringConversions.h`, the call
to `to_string()` will convert the `WCHAR[260]` type to a
`AZStd::wstring_view`, and call the first overload. But if
`StringConversions.h` _has_ been included, the implicit conversion of
`WCHAR[260]` to `bool` has a higher precedence, and it will be chosen
instead.

This overload was causing some uses of `to_string` in
`AnimGraph/GameController.cpp` to resolve to the wrong overload in unity
builds. I have a commit that I made this morning to remove it completely.

**Review by burelc-amzn — APPROVED** [review]

File: Gems/EMotionFX/Code/MCore/Source/StringConversions.h (hunk 9 above)
Comment (lumberyard-employee-dm) [reply Gems/EMotionFX/Code/MCore/Source/StringConversions.h:43]: We can move this to AzCore/Math as well.
We just need to use a constraint that the `bool` overload accepts a type that is exactly `bool` once cv ref qualifiers have been removed
https://godbolt.org/z/G3ce383T1
```
template<class BoolType>
auto to_string(std::string& dest, BoolType value) ->
    std::enable_if_t<std::same_as<std::remove_cvref_t<BoolType>, bool>>
```

**Review by lumberyard-employee-dm — CHANGES_REQUESTED** [review]

**Review by lsemp3d — COMMENTED** [review]
Where these functions unused in EmotionFX? I don't see any updates to the EMFX codebase to use the new AzCore include path and namespace

File: Gems/EMotionFX/Code/MCore/Source/StringConversions.h (hunk 9 above)
Comment (AMZN-puvvadar) [reply Gems/EMotionFX/Code/MCore/Source/StringConversions.h:43]: I'm going to pull the to_string changes out into a separate PR. Now that I'm aware of AzCore/Math/ToString, I think it's warranted.

**Comment by AMZN-puvvadar** [comment]
> Were these functions unused in EmotionFX? I don't see any updates to the EMFX codebase to use the new AzCore include path and namespace

The to_string functions? EMFX was namespacing them under AZStd so the change to include MathStringConversion.h cleanly replaced any usages.

**Comment by AMZN-puvvadar** [comment]
to_string changes are now represented in https://github.com/o3de/o3de/pull/7609 

If and when that change is approved and merged, I'll pull it here to remove those changes from this review. Thanks for the input everyone.

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.cpp [hunk 22]
```diff
@@ -37,7 +37,7 @@ namespace Multiplayer
         return m_filter;
     }
 
-    void MultiplayerDebugAuditTrail::SetAuditTrailFilter(const AZStd::string_view& filter)
+    void MultiplayerDebugAuditTrail::SetAuditTrailFilter(const AZStd::string filter)
     {
         m_filter = filter;
     }
```
Comment (lumberyard-employee-dm) [inline Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.cpp:43]: First this parameter shouldn't be const.

Second the filter string should be moved into m_filter.

By making the parameter const it inhibits moving of the filter parameter into m_filter
```suggestion
    void MultiplayerDebugAuditTrail::SetAuditTrailFilter(AZStd::string filter)
    {
        m_filter = AZStd::move(filter);
    }
```

**Review by lumberyard-employee-dm — CHANGES_REQUESTED** [review]

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.cpp (hunk 22 above)
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.cpp:43]: I hadn't thought of move here, I'll make that change.

**Comment by AMZN-puvvadar** [comment]
Changes made in https://github.com/o3de/o3de/pull/7609 have been merged and resolved

File: Gems/EMotionFX/Code/MCore/Source/StringConversions.h [hunk 23]
```diff
@@ -35,4 +35,4 @@ namespace MCore
 
         static const char* wordSeparators;
     };
-}
\ No newline at end of file
+}
```
Comment (lmbr-pip) [inline Gems/EMotionFX/Code/MCore/Source/StringConversions.h:38]: nit: appears to be just Whitespace change. Consider reverting

File: Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h [hunk 24]
```diff
@@ -7,9 +7,70 @@
...
+    //! Categories for Auditing logs used in MultiplayerDebugAuditTrail
+    enum class AuditCategory
+    {
+        Desync,
+        Input,
+        Event
+    };
+
+    //! @class IMultiplayerAuditingDatum
+    //! @brief IMultiplayerAuditingDatum provides an interface for auditing datums of a given auditing event
```
Comment (lmbr-pip) [inline Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h:23]: nit: "provides an interface for auditing datums of a given auditing event" seems like too many uses of word "auditing"  - it may be accurate, worth a quick review if this is really the responsibilities of this class.

**Review by AMZN-Olex — APPROVED** [review]
Looks good!

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.cpp [hunk 25]
```diff
@@ -0,0 +1,251 @@
...
+
+namespace Multiplayer
+{
+    MultiplayerDebugAuditTrail::MultiplayerDebugAuditTrail()
+        : m_updateDebugOverlay([this]() { UpdateDebugOverlay(); }, AZ::Name("UpdateAuditTrail"))
+    {
+        m_updateDebugOverlay.Enqueue(AZ::TimeMs{ 0 }, true);
+    }
+
+    MultiplayerDebugAuditTrail::~MultiplayerDebugAuditTrail()
```
Comment (lmbr-pip) [inline Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.cpp:30]: nit: Move to default destructor?

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.cpp [hunk 26]
```diff
@@ -0,0 +1,251 @@
...
+    {
+        return m_filter;
+    }
+
+    void MultiplayerDebugAuditTrail::SetAuditTrailFilter(AZStd::string filter)
+    {
+        m_filter = AZStd::move(filter);
+    }
+
+    // --------------------------------------------------------------------------------------------
```
Comment (lmbr-pip) [inline Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.cpp:45]: nit: Not sure why the horizontal line comment here? What are we conveying?

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.cpp [hunk 27]
```diff
@@ -0,0 +1,251 @@
...
+
+        const AZ::u32 stateBefore = m_debugDisplay->GetState();
+        m_debugDisplay->SetColor(AZ::Colors::White);
+
+        m_debugDisplay->SetState(stateBefore);
+    }
+
+    bool MultiplayerDebugAuditTrail::CanPumpAuditTrail()
+    {
+        bool canPump = m_canPumpTrail;
```
Comment (lmbr-pip) [inline Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.cpp:247]: nit: Not too sure why this changes state. Can<I do something> seems like it should not change internal state.

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.h [hunk 28]
```diff
@@ -0,0 +1,75 @@
...
+#include <AzCore/Component/EntityBus.h>
+#include <AzCore/EBus/ScheduledEvent.h>
+#include <AzFramework/Entity/EntityDebugDisplayBus.h>
+#include <Multiplayer/Components/NetworkHierarchyRootComponent.h>
+
+namespace Multiplayer
+{
+    struct AuditTrailInput
+    {
+        AuditTrailInput(AuditCategory category, ClientInputId inputId, HostFrameId hostFrameId,
```
Comment (lmbr-pip) [inline Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.h:20]: nit: should any of these be const ref, rather than passed by value?

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.h [hunk 29]
```diff
@@ -0,0 +1,75 @@
...
+
+        AuditCategory m_category;
+        ClientInputId m_inputId;
+        HostFrameId m_hostFrameId;
+        AZStd::string m_name;
+        AZStd::vector<MultiplayerAuditingElement> m_children;
+    };
+
+    //! Buffer size for ImGui Search bar in Audit Trail UI
+    const int AUDIT_SEARCH_BUFFER_SIZE = 1024;
```
Comment (lmbr-pip) [inline Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.h:38]: nit: Considering adding comments about how set, or implications for changing value.

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.h [hunk 30]
```diff
@@ -0,0 +1,75 @@
...
+        void OnImGuiUpdate(const AZStd::deque<AuditTrailInput>& auditTrailElems);
+
+        //! Draws hierarchy information over hierarchy root entities.
+        void UpdateDebugOverlay();
+
+        //! Checks if the audit trail can be pumped and resets the pump flag
+        bool CanPumpAuditTrail();
+
+        //! Gets string filter for the audit trail
+        AZStd::string_view GetAuditTrialFilter();
```
Comment (lmbr-pip) [inline Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.h:63]: nit: Should any of these be const methods?

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugSystemComponent.cpp [hunk 31]
```diff
@@ -440,67 +477,238 @@ namespace Multiplayer
...
+        m_filteredAuditTrail.clear();
+
+        if (filter.size() == 0)
+        {
+            return;
+        }
+
+        for (auto elem = m_committedAuditTrail.begin(); elem != m_committedAuditTrail.end(); ++elem)
+        {
+            const char* nodeTitle = elem->m_category == AuditCategory::Desync
```
Comment (lmbr-pip) [inline Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugSystemComponent.cpp:629]: Nit: consider using params make nested ternarys easier to read / or split up to aid maintainers.

File: Gems/Multiplayer/Code/Source/NetworkInput/NetworkInput.cpp [hunk 32]
```diff
@@ -86,6 +86,31 @@ namespace Multiplayer
         return m_hostBlendFactor;
     }
 
+    AZStd::vector<MultiplayerAuditingElement> NetworkInput::GetComponentInputDeltaLogs() const
+    {
+        AZStd::vector<MultiplayerAuditingElement> logs;
+        for (uint16_t i = 0; i < m_componentInputs.size(); ++i)
```
Comment (lmbr-pip) [inline Gems/Multiplayer/Code/Source/NetworkInput/NetworkInput.cpp:92]: nit: avoid 'i' as var name

File: Gems/Multiplayer/Code/Source/NetworkInput/NetworkInput.cpp [hunk 33]
```diff
@@ -86,6 +86,31 @@ namespace Multiplayer
...
+    }
+
+    const AZStd::string NetworkInput::GetOwnerName() const
+    {
+        const AZ::Entity* owner = m_owner.GetEntity();
+        if (owner != nullptr)
+        {
+            return owner->GetName();
+        }
+        return "null owner";
```
Comment (lmbr-pip) [inline Gems/Multiplayer/Code/Source/NetworkInput/NetworkInput.cpp:111]: nit: Reason for a specific string rather than empty string?

**Review by lmbr-pip — COMMENTED** [review]
Have some structs and classes with public member variables, consider if you really need to break encapsulation here.

Have a number of class methods that look like they could be const methods in interfaces.

Looks good overall. Some minor comments, leave to your discretion about answering.
